contract/src/lib.rs

```diff
@@ -540,6 +565,136 @@ impl BtcLightClient {
         );
     }
 
+    #[cfg(feature = "zcash")]
+    fn zcash_check_pow_and_equihash(
+        &self,
+        block_header: &Header,
+        prev_block_header: &ExtendedHeader,
+    ) {
+        let config = self.get_config();
+
+        if let Some(pow_allow_min_difficulty_blocks_after_height) =
+            config.pow_allow_min_difficulty_blocks_after_height
+        {
+            // Special difficulty rule for testnet:
+            // If the new block's timestamp is more than 6 * block interval minutes
+            // then allow mining of a min-difficulty block.
+            if prev_block_header.block_height >= pow_allow_min_difficulty_blocks_after_height {
+                if i64::from(block_header.time)
+                    > i64::from(prev_block_header.block_header.time)
+                        + config.pow_target_spacing(prev_block_header.block_height + 1) * 6
+                {
+                    let expected_bits = config.proof_of_work_limit_bits;
+                    require!(
+                        expected_bits == block_header.bits,
+                        format!(
+                            "Error: Incorrect target. Expected bits: {:?}, Actual bits: {:?}",
+                            expected_bits, block_header.bits
+                        )
+                    );
+                }
+            }
+        }
+
+        // Find the first block in the averaging interval
+        // and the median time past for the first and last blocks in the interval
+        let mut current_header = prev_block_header.clone();
+        const MEDIAN_TIME_SPAN: usize = 11;
+        let mut total_target = U256::ZERO;
+        let mut median_time = [0u32; MEDIAN_TIME_SPAN];
+
+        let prev_block_median_time_past = {
+            for i in 0..usize::try_from(config.pow_averaging_window).unwrap() {
+                if i < MEDIAN_TIME_SPAN {
+                    median_time[i] = current_header.block_header.time;
+                }
+
+                let (sum, overflow) = total_target
+                    .overflowing_add(target_from_bits(current_header.block_header.bits));
+                require!(!overflow, "Addition of U256 values overflowed");
+                total_target = sum;
+
+                current_header = match self
+                    .headers_pool
+                    .get(&current_header.block_header.prev_block_hash)
+                {
+                    Some(item) => item,
+                    None => return, // TODO: improve this check
```

Comment:
The somehow should be improved

Reply:
Yes, this silently passes the PoW check if a relevant block header is not available. I don't know under what circumstances that can occur in your code, but it seems bad. (For the case where you reach the genesis block, there is a well-defined behaviour that you should implement if you want to be able to verify the whole chain.)

Reply:
Yeah, the problem is that the light client will be initialized from the arbitrary block, not from the genesis. So we have multiple options here:
- Pass the 22 blocks to the init function here https://github.com/Near-One/btc-light-client-contract/blob/963da1d05b5815309eb844a399c13df15bf0aadb/contract/src/lib.rs#L110
- Ignore this problem, and assume that the trusted relayer submits the first 22 blocks.
- Restrict submitting the first 22 blocks by a trusted relayer

CC @olga24912

Reply:
We need to send the initial blocks first. I think I had a similar issue with Dogecoin, but I don’t remember exactly. I’ll go check how I solved it there.

Reply:
The needed blocks are now submitted on init.
Fixed in 64ca2091a6384b84739a6a65fc66414b80a3a204